File: Sensor Interfaces/SerialInterfaceParser.cs
using System.Text;

namespace QSenseExamples
{
    public static class SerialInterfaceParser
    {
        private const char PREFIX           = '$';
        private const char POSTFIX          = '\n';
        private const char OPCODE_TRANSMIT  = 'T';
        private const char OPCODE_RECEIVE   = 'R';
        private const char OPCODE_UPGRADE   = 'U';

        public static string ByteArrayToString(byte[] data)
        {
            string hexString = "";
            foreach (byte b in data) hexString += b.ToString("X2");
            return hexString;
        }

        public static string CreateTransmitPacket(byte[] coreInterfacePacket)
        {
            string hexString = ByteArrayToString(coreInterfacePacket);
            return "" + PREFIX + OPCODE_TRANSMIT + hexString + POSTFIX;
        }

        public static string CreateUpgradePacket()
        {
            return "" + PREFIX + OPCODE_UPGRADE + POSTFIX;
        }

        private static void ParsePacket(string packet)
        {
            Int32 packetIndex = 0;
            char opcode = '\0';
            byte hi_nible = 0;
            List<byte> dataBuffer = new List<byte>();
            foreach (char c in packet)
            {
                if (PREFIX == c) packetIndex = 0;
                else packetIndex++;

                if (packetIndex == 1)
                {
                    opcode = c;
                    dataBuffer.Clear();
                }

                if (packetIndex > 1)
                {
                    if (POSTFIX == c)
                    {
                        switch (opcode)
                        {
                            case OPCODE_RECEIVE:
                                string packetInfo = "Receive Packet\r\n";
                                packetInfo += $"Handle: {dataBuffer[0]}\r\nData: ";
                                for (int i = 1; i < dataBuffer.Count; i++) packetInfo += dataBuffer[i].ToString("X2");
                                packetInfo += "\r\n";
#if DEBUG
            System.Diagnostics.Debug.WriteLine(packetInfo);
#else
                                Console.WriteLine(packetInfo);
#endif
                                byte[] coreInterfacePacket = new byte[dataBuffer.Count - 1];
                                Array.Copy(dataBuffer.ToArray(), 1, coreInterfacePacket, 0, dataBuffer.Count - 1);
                                QSenseExamples.CoreInterfaceParser.ParsePacket(coreInterfacePacket);
                                break;
                            default:
                                break;
                        }

                        packetIndex = 0;
                    }
                    else if (ValidHex(c))
                    {
                        if (packetIndex % 2 == 0)
                        {
                            hi_nible = (byte)c;
                        }
                        else
                        {
                            dataBuffer.Add(GetValFromHexChars(hi_nible, (byte)c));
                        }
                    }
                    else
                    {
                        packetIndex--;
                    }
                }
            }
        }
        private static bool ValidHex(char c)
        {
            return (c >= '0' && c <= '9') || (c >= 'A' && c <= 'F') || (c >= 'a' && c <= 'f');
        }

        private static byte GetValFromHexChars(byte hi_nible, byte lo_nible)
        {
            Int32 result = 0;
            if (hi_nible >= '0' && hi_nible <= '9') result = hi_nible - '0';
            else if (hi_nible >= 'A' && hi_nible <= 'F') result = 10 + hi_nible - 'A';
            else if (hi_nible >= 'a' && hi_nible <= 'f') result = 10 + hi_nible - 'a';
            result = result << 4;
            if (lo_nible >= '0' && lo_nible <= '9') result += lo_nible - '0';
            else if (lo_nible >= 'A' && lo_nible <= 'F') result += 10 + lo_nible - 'A';
            else if (lo_nible >= 'a' && lo_nible <= 'f') result += 10 + lo_nible - 'a';
            return (byte)result;
        }
    }
}

File: Python Examples/SerialInterfaceParser.py
import CoreInterfaceParser
class SerialInterfaceParser:
    PREFIX = '$'
    POSTFIX = '\n'
    OPCODE_TRANSMIT = 'T'
    OPCODE_RECEIVE = 'R'
    OPCODE_UPGRADE = 'U'

    @staticmethod
    def byte_array_to_string(data: bytes) -> str:
        hex_string = ""
        for b in data:
            hex_string += f"{b:02X}"
        return hex_string

    @staticmethod
    def create_transmit_packet(core_interface_packet: bytes) -> str:
        hex_string = SerialInterfaceParser.byte_array_to_string(core_interface_packet)
        return f"{SerialInterfaceParser.PREFIX}{SerialInterfaceParser.OPCODE_TRANSMIT}{hex_string}{SerialInterfaceParser.POSTFIX}"

    @staticmethod
    def create_upgrade_packet() -> str:
        return f"{SerialInterfaceParser.PREFIX}{SerialInterfaceParser.OPCODE_UPGRADE}{SerialInterfaceParser.POSTFIX}"

    @staticmethod
    def parse_packet(packet: str):
        packet_index = 0
        opcode = '\0'
        hi_nible = 0
        data_buffer = []

        for c in packet:
            if SerialInterfaceParser.PREFIX == c:
                packet_index = 0
            else:
                packet_index += 1

            if packet_index == 1:
                opcode = c
                data_buffer.clear()

            if packet_index > 1:
                if SerialInterfaceParser.POSTFIX == c:
                    if opcode == SerialInterfaceParser.OPCODE_RECEIVE:
                        packet_info = "Receive Packet\r\n"
                        packet_info += f"Handle: {data_buffer[0]}\r\nData: "
                        for i in range(1, len(data_buffer)):
                            packet_info += f"{data_buffer[i]:02X}"
                        packet_info += "\r\n"
                        print(packet_info)

                        core_interface_packet = bytes(data_buffer[1:])
                        CoreInterfaceParser.parse_packet(core_interface_packet)
                    packet_index = 0
                elif SerialInterfaceParser.valid_hex(c):
                    if packet_index % 2 == 0:
                        hi_nible = ord(c)
                    else:
                        data_buffer.append(
                            SerialInterfaceParser.get_val_from_hex_chars(hi_nible, ord(c))
                        )
                else:
                    packet_index -= 1

    @staticmethod
    def valid_hex(c: str) -> bool:
        return ('0' <= c <= '9') or ('A' <= c <= 'F') or ('a' <= c <= 'f')

    @staticmethod
    def get_val_from_hex_chars(hi_nible: int, lo_nible: int) -> int:
        result = 0
        if ord('0') <= hi_nible <= ord('9'):
            result = hi_nible - ord('0')
        elif ord('A') <= hi_nible <= ord('F'):
            result = 10 + hi_nible - ord('A')
        elif ord('a') <= hi_nible <= ord('f'):
            result = 10 + hi_nible - ord('a')

        result <<= 4

        if ord('0') <= lo_nible <= ord('9'):
            result += lo_nible - ord('0')
        elif ord('A') <= lo_nible <= ord('F'):
            result += 10 + lo_nible - ord('A')
        elif ord('a') <= lo_nible <= ord('f'):
            result += 10 + lo_nible - ord('a')

        return result & 0xFF
